Unit tests fail with OpenSSL versions newer than 1.0.1

./configure
make
make test
give me the following result:

> 60/63 Test #60: io_typed_remote_actor ............   Passed    0.01 sec
>       Start 61: io_unpublish
> 61/63 Test #61: io_unpublish .....................   Passed    0.01 sec
>       Start 62: openssl_authentication
> 62/63 Test #62: openssl_authentication ...........   Passed    0.02 sec
>       Start 63: openssl_dynamic_remote_actor
> 63/63 Test #63: openssl_dynamic_remote_actor .....***Exception: Other  0.48 sec
> 
> 98% tests passed, 1 tests failed out of 63
> 
> Total Test time (real) =   1.17 sec
> 
> The following tests FAILED:
> 	 63 - openssl_dynamic_remote_actor (OTHER_FAULT)
> Errors while running CTest
> make[1]: *** [Makefile:130: test] Error 8
> make[1]: Leaving directory '/home/felix/github/actor-framework/build'
> make: *** [Makefile:7: test] Error 2
>

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -45,6 +45,8 @@ if(CAF_BUILD_STATIC_RUNTIME)
       endif()
     endif()
   endforeach()
+else()
+  set(CAF_BUILD_STATIC_RUNTIME no)
 endif()
 
 # add helper target that simplifies re-running configure
@@ -537,7 +539,7 @@ if(NOT CAF_NO_OPENSSL)
     # Check OpenSSL version >= 1.0.1
     if (OPENSSL_VERSION VERSION_LESS 1.0.1)
       message(STATUS
-              "Disable OpenSSL. Required >= 1.0.1 due to TLSv2 support.")
+              "Disable OpenSSL. Required >= 1.0.1 due to TLSv1.2 support.")
       set(CAF_NO_OPENSSL yes)
     else()
       # Check if openssl headers and library versions match
@@ -552,7 +554,7 @@ if(NOT CAF_NO_OPENSSL)
           }
           return -1;
         }
-      " OPENSSL_CORRECT_VERSION_NUMBER )
+      " OPENSSL_CORRECT_VERSION_NUMBER)
       if (NOT OPENSSL_CORRECT_VERSION_NUMBER)
         message(FATAL_ERROR
           "OpenSSL library version does not match headers")

--- a/libcaf_core/caf/config.hpp
+++ b/libcaf_core/caf/config.hpp
@@ -236,5 +236,4 @@
   throw std::runtime_error(msg)
 #endif // CAF_NO_EXCEPTIONS
 
-
 #endif // CAF_CONFIG_HPP

--- a/libcaf_openssl/caf/openssl/session.hpp
+++ b/libcaf_openssl/caf/openssl/session.hpp
@@ -33,6 +33,11 @@ CAF_POP_WARNINGS
 #include "caf/io/network/native_socket.hpp"
 #include "caf/io/network/default_multiplexer.hpp"
 
+#if OPENSSL_VERSION_NUMBER >= 0x10100000L
+# define CAF_SSL_HAS_SECURITY_LEVEL
+# define CAF_SSL_HAS_NON_VERSIONED_TLS_FUN
+#endif
+
 namespace caf {
 namespace openssl {
 

--- a/libcaf_openssl/src/session.cpp
+++ b/libcaf_openssl/src/session.cpp
@@ -204,7 +204,11 @@ const char* session::openssl_passphrase() {
 
 SSL_CTX* session::create_ssl_context() {
   CAF_BLOCK_SIGPIPE();
+#ifdef CAF_SSL_HAS_NON_VERSIONED_TLS_FUN
+  auto ctx = SSL_CTX_new(TLS_method());
+#else
   auto ctx = SSL_CTX_new(TLSv1_2_method());
+#endif
   if (!ctx)
     raise_ssl_error("cannot create OpenSSL context");
   if (sys_.openssl_manager().authentication_enabled()) {
@@ -240,13 +244,13 @@ SSL_CTX* session::create_ssl_context() {
   } else {
     // No authentication.
     SSL_CTX_set_verify(ctx, SSL_VERIFY_NONE, nullptr);
-    auto ecdh = EC_KEY_new_by_curve_name(NID_secp384r1);
-    if (!ecdh)
-      raise_ssl_error("cannot get ECDH curve");
-    CAF_PUSH_WARNINGS
-    SSL_CTX_set_tmp_ecdh(ctx, ecdh);
-    CAF_POP_WARNINGS
-    if (SSL_CTX_set_cipher_list(ctx, "AECDH-AES256-SHA") != 1)
+    SSL_CTX_set_ecdh_auto(ctx, 1);
+#ifdef CAF_SSL_HAS_SECURITY_LEVEL
+    const char* cipher = "AECDH-AES256-SHA@SECLEVEL=0";
+#else
+    const char* cipher = "AECDH-AES256-SHA";
+#endif
+    if (SSL_CTX_set_cipher_list(ctx, cipher) != 1)
       raise_ssl_error("cannot set anonymous cipher");
   }
   return ctx;

